lib/spack/spack/package.py

```diff
@@ -1065,6 +1073,9 @@ def do_stage(self, mirror_only=False):
         if not self.spec.concrete:
             raise ValueError("Can only stage concrete packages.")
 
+        if not self.has_code:
+            raise InvalidPackageOpError("Can only stage a package with a URL.")
```

Comment:
I'm curious whether this is too prickly.  I think it makes sense to raise when trying to `fetch` a no-code package, but I know @becker33 had some use cases where he might want to do something as part of staging and installing the package.   I think some no-source packages like to install metadata about the installation even if the no-source package itself has no real build.  So, I think it's still useful to allow these packages to have a working directory (where `install()` can do something). 

So, I think we should probably allow the call to `do_stage()` but skip the call to `do_fetch()` when `not self.has_code`.

@becker33: thoughts?

Reply:
Sounds reasonable.  It's good to know more about relevant use cases.  Thanks.

Reply:
FYI. Greg just mentioned that same use case w.r.t. `install`.

Reply:
Removed